### ISSUE TITLE
Comments on v0.5

### ISSUE THREAD

**Issue by philippe-hawi** [Dijji/CustomWindowsProperties#10]
1. For fields that have a checkbox, to show the help text, you have to check the box. Is there an easier way to show the help text without modifying the field? Clicking on the name of the field? Or changing the checkbox to a true/false drop list, and showing the help text when the user open the list? This is just a personal preference. I wouldn't like to change the value of a field just to show its help text.

2. Installed panel: clicking on whitespace to clear selection? Decided not to do that? 

3. ColumnIndexType: help text needs attention at the end.

4. isGroup: help text reads "A property description with IsGroup="true" should **specify a ,** otherwise it isn't a useful property". Specify a what? It is like that in the Microsoft docs. I couldn't tell what they meant to say.

5. isTreeProperty: no explanation in MS docs. Current help text says: Whether this is a tree property. Not sure what this means... If you understand what this field is, could you add a clear description?

6. SearchRawValue: same as above.

7. The editor panel looks crammed vertically. The drop list of EditControl barely fits. I suggest adding some space below the last field, and the user can just scroll down in the editor panel.

8. EditControl: typo in the last line of help. "late" should be "later."

9. Weird issues when saving and deleting. I'll try my best to explain what happened. This was the first property i'm editing. No saved properties yet. I clicked on save. The prop got saved, but a message box appeared asking if i want to discard changes. I chose no. Then i went to the saved panel. Right clicking on the property resulted in the same message. I had to discard changes (there weren't any changes though) to be able to delete it. I tried reproducing this behavior but i couldn't. But then this happened again when i created a new set of properties. Will try to better understand when this is happening. But what i am sure of is that  when this message appears, there definitely aren't any changes to be discarded.

10. I created the following property: Custom.Documents.Journal. Context menu of Journal has the 3 options available. For Documents, only export is available. What if I want to delete (or even install) all properties under Documents? For Custom, nothing is available (again, i might want to install or delete everything in that category). Does it even make sense to export anything that's not in the lowest level of the tree? The exported propdesc of Custom.Documents has information for Custom.Documents.Journal. I don't think Custom.Documents should have a propdesc in the first place.

11. When i select Documents, i expected the editor to show a blank new property with the name Custom.Documents, which can not be saved obviously, where i would add something to the name (e.g., .Year). Is that unreasonable or not an expected behavior? Anyway, after i create Year and save it (now i have in the saved list Custom.Documents.Journal and Custom.Documents.Year), if i click on Year then on Documents, the editor displays the details of Year. If i click on Journal then Documents, the editor displays the details of Journal. I think there's something to be addressed there.

12. Having created Custom.Documents.Journal and installed it (now the install button is disabled), i create Custom.Videos and save it. The install button is enabled. I then click on Custom.Documents and the install button is still enabled. I then click on Custom.Documents.Journal, the install button gets disabled, and stays disabled when i select Custom.Documents. So in essence, moving from a non-installed property to an installed property disables the install button (correct), but moving from a non-installed property to the group header of an installed property does not disable the install button (incorrect, given the way things are right now, which is the installed button always disabled for group headers; i argue that if there are properties yet to be installed, the user should be able to install them all through their group header).

13. Still no indication of which saved properties are installed in the saved panel. Is that something you decided not to do?

This is where I stopped testing. I have deadlines but I'll keep testing whenever I can and report back. Hopefully, everything will be thoroughly tested by end of next week.

**Comment by Dijji** [comment]
Excellent feedback! Thank you.

I have had great fun going through all your comments, and have now released version 0.6 incorporating the resulting changes. Please use this new version as the basis for your further testing.

Here are my detailed point by point responses:

- 1: Clicking on the area to the left of a control in the editor now shows its help, and selects the control quietly, without changing its value.
- 2: It is true that there is no way back to the original state of having no installed property selected once one has been selected. What I don't understand is why this matters. Clearing the selection gains you nothing, but it would remove the option of copying install property values to the editor, or of comparing them to the current editor values. Why do it?
- 3: Help text fixed (3, 4, 8)
- 5: Speculation added to help text for mystery values (5, 6). But in fact, I know, no more than you do. Experimentation may elucidate matters.
- 7: I actually tried out a couple of ways of spacing the controls out a little more, but they made things worse, so I abandoned them. My starting point was the combo boxes near the bottom, as they are the most crowded. The other combo boxes follow naturally. However, they are already the tallest controls, and this makes them taller. To make them less out of place, this leads to spreading out the text boxes. By now, the spacing of the names in the left column is becoming rather uneven, making them difficult to scan down looking for a property. Also, the checkboxes now look like they come in clumps, so they need to be spread out as well. Now the labels are a little bit more even, but still hard to scan because they are becoming further and further apart. At which point I revert to the original layout, and greet it with a sigh of relief. Much better. My conclusion: that the single highest priority is to make scanning down the list of property names as easy as possible. On the EditControl combo: it is not the Windows convention to reserve space to expand combo boxes. In fact, they are allowed to expand even outside the frame of the window in which they appear. So I'm leaving this as is.
- 9: I believe you on this one. I tightened up the code to increase the frequency of checks for editor dirtiness and immediately started to run in to the sort of problems you report. So I have cleaned up the code in this area, and hope to have eliminated the problem.
- 10: This is definitely a case where a group action on the properties make sense. The scenario is where somebody has created a set of custom properties for their software, and for package installation purposes, wants to define all of their custom properties in a single .propdesc file, which is the standard practice (Office distribute their custom properties in this way, for example). Where I fell short is enabling you to export from any level, including all properties in the chosen subtree. This has now been remedied. There is still the need to create a command line utility to install the propdesc when invoked by the installation package, which is on the work list. This would also come in handy for you, should you want a quick way of copying your custom properties to another machine.
- 11: I think creating a new potential child property is a really good idea. Done.
- 12: 11 fixes most of the problems in 12. I have also finally conceded, having thought about your usage scenarios and found them convincing, and implemented support for commands on parent nodes: delete and install in the saved tree; and uninstall in the installed tree. In each case, the command is performed on all of the eligible children in the subtree under the selected node. Because users may sometimes be surprised by what will and won't be operated on, the user is told what the command targets will be, and asked to confirm the operation.
- 13: This was implemented a version or so back. The background of properties in the saved tree that have also been installed should be green. However, you may not have seen it because there was a bug where if a property has a DisplayName, then the highlight did not appear in the saved tree after the property was installed until the whole app was restarted (strange but true!).

Dijji

**Comment by philippe-hawi** [comment]
Follow-up:

2: I found myself multiple times try to clear the middle bottom panel. I think it's just muscle memory for me? In windows, say i'm in a media folder. I click on a file to see its properties in the details panel on the right. If I'm done with that, I intuitively like to clear the selection to "reset" the state of the folder. Same thing here. If I click on a property to check something, or even by mistake, I don't want the panel to be stuck on showing the details of that property. For me, deselecting a property is a kind of refresh to the view to keep the panels clean and show only what the user wants to see and/or is using. The same applies to the saved properties panel. Actually, if I want to start a new property with default values after having selected a saved property, there's no way to do that. I just think the user should have the ability to restore a clean refreshed view of the panels as they are when you first open the program. But, again, it's a matter of preference and aesthetics that doesn't affect the functionality (too much), so I won't be mentioning it again.

7: Got it. Thinking about something else now. Not sure if it's worth it, or even easy enough to do: resizable panels? Just throwing it out there. Haven't even thought of a need for it but it's just something i see in this type of GUIs.

10: This makes sense now, and i like it.

11 and 12: Just perfect!

13: That's right. However: when i select an installed property, the font turns white. It's barely visible against the green background. Is it just me? I'll have to show this to someone else to get feedback.

Everything sounds great. I'll continue testing on v0.6.

**Comment by Dijji** [comment]
2: in fact, you can now clear the middle bottom panel by selecting any of the parent nodes in the installed tree. It's a side effect of enabling their context menus. And if you want to start a new property with default values, click the New button, which gets the editor exactly to the place it is at when the app starts. Do these things help?
7: Between the titles at the top of the window are slightly thicker bars. These are actually resize handles which will let you rebalance the panels left to right, at least.
13: I will take a look. Changing the colour is not difficult.

Note that the new button and the resize handles are both documented in the README on the front page of the website, which is regularly updated. You might find interesting to read back through it from time to time to see if there is anything new that catches your fancy.

I look forward to the feedback on 0.6!

Dijji

**Comment by Dijji** [comment]
Here's what's going on with the colours. Selecting an item in the tree does two things: changes background colour to dark blue; and changes font colour to white. Installing a property also changes the background colour, this time to light green. When you select an installed property, the background colour comes from it being installed, but the font from it being selected. Hence white on light green, which is not very legible.

It is actually difficult to find a background colour on which both black and white fonts are legible.

A better solution is to change the background colour used for selection so that the font can remain black, and avoid the conflict.

Here is a selection of the parent of an installed property:

![clip1](https://user-images.githubusercontent.com/27084934/81609859-7f831880-93d0-11ea-9bd3-c05088d44325.PNG)

The selection is marked with light blue. The next picture shows the selection of an installed property. The background remains light green, but fortunately, selection also changes the font to bold, so the selection remains visible.

![clip2](https://user-images.githubusercontent.com/27084934/81609737-4d71b680-93d0-11ea-9d57-ce2e5797a0dc.PNG)

Does this work for you?

Dijji
click

**Comment by philippe-hawi** [comment]
Yes! This is much better. I think you should definitely do this.

**Comment by philippe-hawi** [comment]
> 2: in fact, you can now clear the middle bottom panel by selecting any of the parent nodes in the installed tree. It's a side effect of enabling their context menus. And if you want to start a new property with default values, click the New button, which gets the editor exactly to the place it is at when the app starts. Do these things help?
> 7: Between the titles at the top of the window are slightly thicker bars. These are actually resize handles which will let you rebalance the panels left to right, at least.
> 13: I will take a look. Changing the colour is not difficult.
> 
> Note that the new button and the resize handles are both documented in the README on the front page of the website, which is regularly updated. You might find interesting to read back through it from time to time to see if there is anything new that catches your fancy.
> 
> I look forward to the feedback on 0.6!
> 
> Dijji

2. Yes. They have the same end result, which is good enough for me and my use.

7. Should've clarified that I was actually thinking up/down, in case either the saved list or the installed is too long. Left/right, there's plenty of room already. But come to think about it, i think the current layout is perfect and i wouldn't resize anything anyway. If the list is long, i would just scroll down.

Will take another look at updated README!

**Comment by Dijji** [comment]
I have tweaked 0.6 to implement the colour changes to create 0.6.1 and updated the release with it, so you can try the changes out in practice.

I did look at the up down split a while back, but there is nowhere obvious to split the screen into upper and lower halves, as the central bar looks both ways. I could have split the bar into two, but only at the cost of valuable space. Anyway, as you say, it's not essential, with scrolling available.

**Comment by Dijji** [comment]
WARNING: If you have been trying to test with using the properties in File Meta, as I have, then you will probably have come across this bug, which I found and killed this morning.

- Properties were being assigned a PropertyId value of 1 if they were the first property under a particular parent. But 1 is a reserved value. Installation works, but the value of such a property cannot be set. The lowest legal value is 2. This has now been fixed for new properties, and values of 1 will be corrected when existing properties are saved. Installed properties with this problem will need to be uninstalled, saved and then reinstalled.

Let me know if you would like an early release with this fix.

**Comment by Dijji** [comment]
I have now released version 0.7 containing fixes all the bugs that my testing has found this week.

Please use this version for any further testing.

**Comment by Dijji** [comment]
Here are today's fixes. Nothing big enough to warrant a new release, but just make you aware:
- SortDescription: SmallestBiggest should have been SmallestLargest
- Installing with differences to Installed radio button checked generates spurious loss of changes Message Box.
- Correct help text for GroupingRange (error was copied from Microsoft documentation)

**Comment by Dijji** [comment]
Today I have been testing Windows Explorer search working with custom properties. No further bugs. The results on searches:
- Searching for values works fine
- Searching prefixed by terms does not seem to happen. For instance, I was hoping that if you specified a Mnemonic of mmm then you would be able to use a search like mmm:value , but no such luck so far.

**Comment by philippe-hawi** [comment]
I've been out for a while. Too many deadlines. I'll resume testing today (hopefully). Just downloaded v0.7 (Chrome flagged it as dangerous though; first time this happens).
About the Explorer search, I tried it in the early versions. Wasn't getting any results so I dropped it since I don't even use the Explorer search. I'll try testing it further anyway.

**Comment by philippe-hawi** [comment]
- If I click on a saved property, its details appear in the editor. If I click on "New", the editor is blank. Now if I click on a saved property, the editor doesn't update. If I right-click on a saved property, the editor updates.

- When I save a new property, the list of saved properties collapses. Can you keep it expanded?

- I'm not sure what's happening here, but some properties are missing fields?  When you start a new property, "IsColumnSparse" and "ColumnIndexType" are not there, but older properties have these fields.

- For "Basics" -> "Name": clicking on Name doesn't make the help text appear. All other fields work correctly.

**Comment by Dijji** [comment]
On your points:
- I cannot reproduce this. Maybe I have fixed it accidentally. Could you try again on the next version?
- The collapse happens only if the new property starts with something new, and so adds a root to the tree. I will look at stopping this.
- This is because the two properties you mention are not legally settable, unless IsColumn is true. If you set IsColumn on, they will reappear.
- You are right. I will fix this.

More generally, it looks as though we are about done. My only issue is that I still have not got Mnemonics working. I would have thought that for your needs, being able to make a query in Explorer like "myp: value" where myp is a mnemonic for Custom.MyProperty is important. Even worse, "Custom.MyProperty: value" doesn't work either. All I can get working for the moment is the straight search for "value".

How significant is this for you?

Dijji

**Comment by philippe-hawi** [comment]
There multiple angles for my use of custom properties. The main one is organizing and sorting file. But another, less important but important nonetheless, is searching the way you described. If it doesn't work, it's not the end of the world; I'll find a workaround somehow... But I do hope this works down the line. Also, I think getting only "Custom.MyProperty: value" to work is not very valuable. I don't think anyone would search with the full name instead of the mnemonic.
Maybe it's an indexing issue? There are launchers or search tools that can search through file metadata (I use Listary, but it only searches filenames, not metadata; I've heard about Everything, Wox... not sure if they have that functionality). Perhaps try one of those and see if it can search for the custom properties?

About my points:

- OK. Again, what I do is: click on a saved property, click on new, click on that same saved property. The editor stays blank.

- Also, if I add a property using an existing top level, the properties of other top level collapse. E.g. say I have save Custom1.Prop1 and Custom2.Prop1, then I add Custom2.Prop2: Custom1 collapses.

- Got it. Thanks.

**Comment by Dijji** [comment]
Okay, I have posted release 0.8 containing all the fixes. On your issues:
- Ah, it had to be the *same* saved property. Fixed.
- The saved tree should now remain always fully expanded.
- You can now click on the label of the canonical name and be helped.

Could you please check that the fixes are good?

Generally, this really does feel about it. If you pass the fixes, my only holdout will be trying to get mnemonics working for a little while longer. I'm currently working with a Windows 7 test machine to see if things used to be any different. But if I make no progress, then I'll vote for declaring the first version done. Does the outcome meet the needs you came in with?

**Comment by philippe-hawi** [comment]
I had some properties saved. Created a new one, A.B, and saved it. In the tree, A.B is expanded, and so are other properties, but only up to that level. Everything after level 2 got collapsed.
No other comments.
If mnemonics work, great. If not, you have your v1.
The program does indeed meet everything I thought about. It's actually much better than I thought such a software could be.

**Comment by Dijji** [comment]
Well, I'm pleased that you are pleased with it! I do have to say, however, that it would not have been nearly as good without your input. I think that the way we have worked together is classic proof of the benefits of getting user input in in the early stages, while the program is still plastic. Thank you!

On the collapsing saved tree, you are exactly right. This issue has been a bit like a rodent infestation: just when you think you have blocked up the last hole... Still, I think I have it this time. It will be in the next release, which will be 1.0 (after a little bit more sweat applied to the question of mnemonics, whether I succeed or not).

Dijji

**Comment by philippe-hawi** [comment]
Totally agree! It has been fun testing, talking, and watching the program develop.
After using your FileMeta, I was hoping you would write something of similar quality for custom properties, and this program certainly did not disappoint. Appreciate your hard work.
Looking forward to 1.0 👍

**Comment by philippe-hawi** [comment]
Hello. Just wondering if this is still alive!

**Comment by Dijji** [comment]
Yes, it is still alive. It is just that I was getting nowhere with getting mnemonics to work, so rather than continue to flog a pretty dead horse, and believing that you in particular had a usable tool, I turned my attention to one of my current research activities, which has turned out to be much more rewarding. However, I haven't quite given up, it's just waiting until the mood strikes me to do another bit of head battering!

**Comment by philippe-hawi** [comment]
Got it.
For my use, what I have is indeed enough. Just wanted to check in :)
Best.
